### PR TITLE
Improved parser

### PR DESCRIPTION
**Pull request by apfohl**
- Implemented strdup based on GC_malloc
- Created parser rules for parsing G-code headers

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -8,10 +8,10 @@ YFLAGS = -q
 LEX = flex
 LFLAGS = --header-file=lexer.h
 
-SOURCES = parser.c lexer.c g-compiler.c
+SOURCES = parser.c lexer.c memory.c g-compiler.c
 OBJECTS = $(patsubst %.c, %.o, $(SOURCES))
 
-.PHONY: release lemon clean
```
Comment (apfohl) [inline Makefile:14]: This makes the Makefile work as expected. If already compiled do not recompile.

File: g-compiler.c [hunk 2]
```diff
@@ -58,11 +59,11 @@ int main(int argc, char **argv)
 
     int token;
     while ((token = yylex(lexer))) {
-        if (token == -1) {
```
Comment (apfohl) [inline g-compiler.c:61]: https://en.m.wikipedia.org/wiki/Yoda_conditions

File: g-compiler.c [hunk 3]
```diff
@@ -58,11 +59,11 @@ int main(int argc, char **argv)
...
+        if (-1 == token) {
             break;
         }
 
-        char *text = yyget_text(lexer);
```
Comment (apfohl) [inline g-compiler.c:65]: This duplicates the string for further usage inside the parser. E.g. safe it as a value.

File: memory.c [hunk 4]
```diff
@@ -0,0 +1,22 @@
+#include <string.h>
+#include <gc/gc.h>
+
+#include "memory.h"
+
+char *gc_strdup(const char *source)
```
Comment (apfohl) [inline memory.c:6]: This function is an implementation of strdup with malloc from LibGC.

File: parser.y [hunk 5]
```diff
@@ -1,13 +1,26 @@
 %include {
     #include <assert.h>
+    #include <stdio.h>
 }
 
-start ::= statement_list.
```
Comment (apfohl) [inline parser.y:5]: In case of a syntax error.

File: parser.y [hunk 6]
```diff
@@ -1,13 +1,26 @@
...
+%syntax_error {
+    fprintf(stderr, "%s", "Syntax Error!\n");
+}
+
+%start_symbol start
+
+start ::= header statement_list.
+start ::= error. {
```
Comment (apfohl) [inline parser.y:13]: If a parsing error occurs.

File: parser.y [hunk 7]
```diff
@@ -1,13 +1,26 @@
...
+    fprintf(stderr, "%s", "Parser Error!\n");
+}
+
+header ::= PERCENT statement_list PERCENT. {
```
Comment (apfohl) [inline parser.y:17]: This makes the header.

Comment (maknoll) [reply parser.y:17]: Ich füge noch ne Regel ein, falls statt "%"  "M48" und "M95" verwendet werden.

Comment (apfohl) [reply parser.y:17]: Is this necessary for the prototype? Can this happen in the files that will be used first?

Comment (maknoll) [reply parser.y:17]: The sample files, which are exported from kicad, use M48 for the header start and % for header end. My merger uses % and %. Its not necessary, i think testing would be easier if it implements all permutations here.

**Pull request merged**